source/common/usm_pool_config.cpp

```diff
@@ -0,0 +1,267 @@
+    Configs[PoolDescriptor(UR_USM_TYPE_SHARED)].Capacity = 0;
+    Configs[PoolDescriptor(UR_USM_TYPE_SHARED)].SlabMinSize = 2_MB;
+
+    // Allow pooling of shared allocations that are only modified on host.
+    Configs[PoolDescriptor(UR_USM_TYPE_SHARED, true)].MaxPoolableSize = 4_MB;
+    Configs[PoolDescriptor(UR_USM_TYPE_SHARED, true)].Capacity = 4;
+    Configs[PoolDescriptor(UR_USM_TYPE_SHARED, true)].SlabMinSize = 2_MB;
+
+    // TODO: replace with UR ENV var parser and avoid creating a copy of
+    // 'config'
+    auto GetValue = [](std::string &Param, size_t Length, size_t &Setting) {
+        size_t Multiplier = 1;
+        if (tolower(Param[Length - 1]) == 'k') {
+            Length--;
+            Multiplier = 1_KB;
+        }
+        if (tolower(Param[Length - 1]) == 'm') {
+            Length--;
+            Multiplier = 1_MB;
+        }
+        if (tolower(Param[Length - 1]) == 'g') {
+            Length--;
+            Multiplier = 1_GB;
+        }
+        std::string TheNumber = Param.substr(0, Length);
+        if (TheNumber.find_first_not_of("0123456789") == std::string::npos) {
+            Setting = std::stoi(TheNumber) * Multiplier;
+        }
+    };
+
+    auto ParamParser = [GetValue](std::string &Params, size_t &Setting,
+                                  bool &ParamWasSet) {
+        bool More;
+        if (Params.size() == 0) {
+            ParamWasSet = false;
+            return false;
+        }
+        size_t Pos = Params.find(',');
+        if (Pos != std::string::npos) {
+            if (Pos > 0) {
+                GetValue(Params, Pos, Setting);
+                ParamWasSet = true;
+            }
+            Params.erase(0, Pos + 1);
+            More = true;
+        } else {
+            GetValue(Params, Params.size(), Setting);
+            ParamWasSet = true;
+            More = false;
+        }
+        return More;
+    };
+
+    auto MemParser = [&Configs,
+                      ParamParser](std::string &Params,
+                                   PoolDescriptor poolDesc =
+                                       PoolDescriptor(UR_USM_TYPE_UNKOWN)) {
+        bool ParamWasSet;
+        PoolDescriptor LM = poolDesc;
+        if (poolDesc.type == UR_USM_TYPE_UNKOWN) {
+            LM = PoolDescriptor(UR_USM_TYPE_HOST);
+        }
+
+        bool More =
+            ParamParser(Params, Configs[LM].MaxPoolableSize, ParamWasSet);
+        if (ParamWasSet && poolDesc.type == UR_USM_TYPE_UNKOWN) {
+            for (auto &Config : Configs) {
+                Config.second.MaxPoolableSize = Configs[LM].MaxPoolableSize;
+            }
+        }
+        if (More) {
+            More = ParamParser(Params, Configs[LM].Capacity, ParamWasSet);
+            if (ParamWasSet && poolDesc.type == UR_USM_TYPE_UNKOWN) {
+                for (auto &Config : Configs) {
+                    Config.second.Capacity = Configs[LM].Capacity;
+                }
+            }
+        }
```

Comment:
I'd like to ask, what was the intended behavior for when invalid type is passed in the string config (like ` foo:0,3,2m`)? Since, as it is now,  the invalid type's parameters are overwriting other config types in the `ParamParser` (found in the test), except for the first parameter- `MaxPoolableSize` because in the `MemTypeParser` the invalid type name is never erased from the string so this parameter isn't returned from the `GetValue`. Perhaps It'd be a good idea to add  some checking to the `ParamParser` if the type isn't valid and erase it and it's parameters from the config string?

Reply:
Yes, I think it's a good idea. We should also log each invalid type and say we're going to use the defaults. There is some code for printing the configuration in this file but I'd like to replace it with our logging mechanism in the future.